### DISCUSSION TITLE
v0.5.0-preview

### DISCUSSION THREAD

**Discussion by progrium** [progrium/darwinkit#203] [Announcements]
### MacDriver is becoming DarwinKit! 

> I think this might now be the best bindings project in existence for Apple APIs. Possibly even the best way to make small utilities on the Mac. And soon even other Apple devices.

This is a preview release for `v0.5.0` which is a major rewrite and expansion on these bindings. This release is incompatible with previous releases, however there is an unmaintained [legacy branch](https://github.com/progrium/macdriver/tree/legacy) and all the [previous releases](https://github.com/progrium/macdriver/releases) for existing code to use. The API in this release is *very* close to what a 1.0 release would look like, so there shouldn't be significant changes only additions to what's here now, which includes:

* Over [30 frameworks](https://pkg.go.dev/github.com/progrium/macdriver/macos@main#section-directories) with nearly complete bindings and more are on the way
* Over __2,000__ classes with near complete coverage. Previously we had *partial* coverage of about 80.
* __22,000__ methods/properties and nearly __10,000__ constants/enums. We had less than 50 constants before.
* Automatic conversion and use of native Go builtin types in APIs like strings, bools, []byte, maps, etc. 
* Support for block arguments as Go functions with properly typed arguments (callbacks!)
* Pre-made delegate implementations you can simply set Go functions on
* More 1-to-1 mapping to Objective-C symbol names while also improving Go idiomaticness
* Documentation for all symbols including a link to official Apple docs on that symbol
* Simple model for working with Objective-C [memory management](https://github.com/progrium/macdriver/blob/main/docs/memorymanagement.md)
* As the new name implies, we are set up to support more platforms like iOS, tvOS, watchOS, maybe visionOS
* Continued support for both Intel and Apple Silicon
* No more compile warnings, nearly all previous GitHub issues are resolved
* [More examples](https://github.com/progrium/macdriver/tree/main/macos/_examples), [better docs](https://pkg.go.dev/github.com/progrium/macdriver@main), and bindings are nearly all [generated](https://github.com/progrium/macdriver/blob/main/docs/generation.md)

However, in this preview release we have a few things left out that with your help we can get into the final `v0.5.0` release:

* We are missing generated bindings for [framework functions](https://github.com/progrium/macdriver/issues/151)
* We are currently covering APIs up to macOS 12 (which is most tbh), but we can easily cover up to 14 [with help](https://github.com/progrium/macdriver/issues/170)
* Our coverage of structs is incomplete, many are [empty placeholders](https://github.com/progrium/macdriver/issues/183)
* [Several](https://github.com/progrium/macdriver/issues/172) [examples](https://github.com/progrium/macdriver/issues/187) are currently broken

**Another way you can help is to let me know what kind of examples you want to see. Drop your requests in [Discussions](https://github.com/progrium/macdriver/discussions).**

This preview release is an official reboot of the project with a new name: DarwinKit. Between now and the final release, the repository will remain `macdriver` while everything else updates to DarwinKit, and when we release `v0.5.0` we'll [rename the repository](https://github.com/progrium/macdriver/issues/184) as well.

A huge thanks to @mkrautz for his work 10 years ago that I found and started this project on in 2020. Unfortunately most of his code is now gone, replaced by the more recent great work by @hsiafan. The previous code generation was by @mgood, which has also been majorly upgraded. Also thanks to @tmc and @programmingkidx for engaging more recently and motivating me to do this rewrite. But also thanks to all the [previous contributors](https://github.com/progrium/macdriver/graphs/contributors) that helped get to this point.

<hr /><em>This discussion was created from the release <a href='https://github.com/progrium/macdriver/releases/tag/v0.5.0-preview'>v0.5.0-preview</a>.</em>

**Comment by TotallyGamerJet** [comment]
What an exciting release. Congrats! Do you think that porting the project to be cgo-less would be beneficial? I maintain [purego](https://github.com/ebitengine/purego) a library designed to call C functions from Go without Cgo. It's used in ebitengine although not entirely yet. I'd be willing to start the work in the generator if this something you'd be interested in.

**Reply by progrium** [reply]
My immediate response is of course, but then "how" and just thinking through the implications. You do it by runtime linking with a shared library? So you need `libobjc.A.dylib` ... for some reason I don't on my system. The fact that the Go binary would not be "self contained" is maybe not an issue if you can trust the library is "always" there, like libc ... but if I don't have it, I'm sure others don't have it. And then I'm curious about stuff outside the basic runtime, which I see you have some experience with. We used to set constants from C, which we don't do anymore, but what about framework functions?

**Reply by TotallyGamerJet** [reply]
Every macOS version has libObjc. That's how DarwinKit works. In your project it's just the C code that does the linking but my library does it without C. The reason it may appear that you don't have it is because macOS now uses a shared cache where all the libraries (frameworks) are stored and so there isn't a single file for each library. You can run the example with CGO_ENABLED=0 and see that it runs with no issues.

**Reply by TotallyGamerJet** [reply]
> And then I'm curious about stuff outside the basic runtime, which I see you have some experience with. We used to set constants from C, which we don't do anymore, but what about framework functions?

I'm not sure exactly what you mean by basic runtime? Any class can be created or called through the objc package. Any C or ObjC function can be called (as long as the types are supported which most are) There is no issue calling any framework function.

**Reply by progrium** [reply]
Huh, yea your objc package tests work for me. I still don't understand how you can reference a file that isn't there, but I'm assuming it's some dynamic linker/loader magic? I do see with `otool` now that the binaries we produce are linking against a `/usr/lib/libobjc.A.dylib`, I had just assumed it was a static library. 

That covers the objective-c runtime, so yes, any method on any class. And then I had the same previous assumption with frameworks, but it looks like they're dynamic too. Huh. 

Alright, so next I'm wondering about libffi, which we use I think to get around cgo issues, and that wouldn't be a thing. You would basically take the place of libffi. What about callbacks? 

Also, why would you need to touch any generation, wouldn't you just need to update our `objc` package to be backed by purego?

**Reply by progrium** [reply]
There is cgo referenced in the generated framework packages, but it's basically no longer used.

**Reply by progrium** [reply]
Actually we do have to [do this](https://github.com/progrium/macdriver/blob/main/macos/appkit/protocols.gen.m), and that's the only reason cgo is needed in frameworks. I don't fully understand why, but without this delegates don't work. I'm not sure what the equivalent would be with purego.

**Reply by hsiafan** [reply]
(Unlike class objects) The compiler creates a protocol object for each protocol declaration it encounters, but only if the protocol is also:
1. Adopted by a class, or
2. Referred to somewhere in source code (using @protocol())

see [ocProtocols](https://developer.apple.com/library/archive/documentation/Cocoa/Conceptual/ObjectiveC/Chapters/ocProtocols.html#//apple_ref/doc/uid/TP30001163-CH15)

The objc package protocol method signatures from protocol objects. If can store/get those metadata  in somewhere go code, the protocols.gen.m is no longer need.

**Reply by TotallyGamerJet** [reply]
> You would basically take the place of libffi. What about callbacks?

Yes purego would replace libffi. Purego supports callbacks so that shouldn't be a problem

> Also, why would you need to touch any generation, wouldn't you just need to update our objc package to be backed by purego?

Oh I just assumed the generator was creating Cgo wrappers for each class. I was mistaken. Yes updating the objc package to use purego would be most of the work.

As for the protocols I'm not sure. We haven't had any issues with them not working so I think just testing to see if it works would be the easiest thing to do.

**Reply by progrium** [reply]
Well it sounds like this could be super rad. And I want to understand how purego works, I don't know why it's not more common. I'm pretty exhausted after this release and I have a ton of other projects, but I'd say if you wanted to port our objc package and get tests passing that would be most of the work. 

Meanwhile, @tmc, maybe instead of generating C/cgo for functions we try purego. This could be independent of a port of objc and gets us used to purego, which I'd probably use everywhere and in other projects if it works out.

**Reply by TotallyGamerJet** [reply]
I'll take a look at porting the objc package. I'm currently working on some purego changes so hopefully in a few weeks I'll have a PR

**Comment by hsiafan** [comment]
You guys did an amazing job, Congratulations! 🎉

**Reply by progrium** [reply]
While you're around, what were the reasons you started using libffi?

**Reply by hsiafan** [reply]
I need a way to invoke C functions and generate C callbacks. libffi is easy to use and fully meets my requirements.
